Lib/test/test_dataclasses/__init__.py

```diff
@@ -1317,6 +1317,29 @@ def __post_init__(self, init_base, init_derived):
         c = C(10, 11, 50, 51)
         self.assertEqual(vars(c), {'x': 21, 'y': 101})
 
+    def test_init_var_name_shadowing(self):
+        # Because dataclasses rely exclusively on `__annotations__` for
+        # handling InitVar and `__annotations__` preserves shadowed definitions,
+        # you can actually shadow an InitVar with a method or property.
+        #
+        # This only works when there is no default value; `dataclasses` uses the
+        # actual name (which will be bound to the shadowing method) for default
+        # values.
+        @dataclass
+        class C:
+            shadowed: InitVar[int]
+            _shadowed: int = field(init=False)
+
+            def __post_init__(self, shadowed):
+                self._shadowed = shadowed
+
+            @property
+            def shadowed(self):
+                return self._shadowed
+
+        c = C(5)
+        self.assertEqual(c.shadowed, 5)
```

Comment:
nit: maybe we could mutate the value that's passed in, to make it clearer for the reader of the test that the value passed into `__init__` could be "intercepted" twice before it's accessed on an instance of `C`:

```suggestion
                self._shadowed = shadowed * 2

            @property
            def shadowed(self):
                return self._shadowed * 3

        c = C(5)
        self.assertEqual(c.shadowed, 30)
```